substrate/frame/system/src/extensions/check_weight.rs

```diff
@@ -130,49 +129,23 @@ where
 	}
 }
 
-/// Check that the combined extrinsic length and proof size together do not exceed the PoV limit.
-pub fn check_combined_proof_size<Call>(
-	info: &DispatchInfoOf<Call>,
-	maximum_weight: &BlockWeights,
-	next_len: u32,
-	next_weight: &crate::ConsumedWeight,
-) -> Result<(), TransactionValidityError>
-where
-	Call: Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>,
-{
-	// This extra check ensures that the extrinsic length does not push the
-	// PoV over the limit.
-	let total_pov_size = next_weight.total().proof_size().saturating_add(next_len as u64);
-	if total_pov_size > maximum_weight.max_block.proof_size() {
-		log::debug!(
-			target: LOG_TARGET,
-			"Extrinsic exceeds total pov size. Still including if mandatory. size: {}kb, limit: {}kb, is_mandatory: {}",
-			total_pov_size as f64/1024.0,
-			maximum_weight.max_block.proof_size() as f64/1024.0,
-			info.class == DispatchClass::Mandatory
-		);
-		return match info.class {
-			// Allow mandatory extrinsics
-			DispatchClass::Mandatory => Ok(()),
-			_ => Err(InvalidTransaction::ExhaustsResources.into()),
-		};
-	}
-	Ok(())
-}
-
 /// Checks if the current extrinsic can fit into the block with respect to block weight limits.
 ///
 /// Upon successes, it returns the new block weight as a `Result`.
 pub fn calculate_consumed_weight<Call>(
 	maximum_weight: &BlockWeights,
 	mut all_weight: crate::ConsumedWeight,
 	info: &DispatchInfoOf<Call>,
+	len: usize,
 ) -> Result<crate::ConsumedWeight, TransactionValidityError>
 where
 	Call: Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>,
 {
-	let extrinsic_weight =
-		info.weight.saturating_add(maximum_weight.get(info.class).base_extrinsic);
+	// Also Consider extrinsic length as proof weight.
+	let extrinsic_weight = info
+		.weight
+		.saturating_add(maximum_weight.get(info.class).base_extrinsic)
+		.saturating_add(Weight::from_parts(0, len as u64));
```

Comment:
Stupid question: Is this also used for inherents? (Does this have any influence on weights on the relaychain as relevant to parachain consensus?)

Reply:
Specifically: We already add lengths to the weight in paras inherent for checking within the paras inherent to not exceed the weight e.g. [here](https://github.com/paritytech/polkadot-sdk/blob/d2fd53645654d3b8e12cbf735b67b93078d70113/polkadot/runtime/parachains/src/paras_inherent/weights.rs#L123). I hope we are not counting it twice now?

Reply:
I don't understand what is the reasoning is for this. Why are we setting the proof size there? On relay chains the maximum proof size is `U64::MAX` AFAIK, effectively disabling any proof size related checks.

Reply:
Good point @skunert . I think we should remove and clean the code in parachain inherent. Proof size is indeed only relevant for parachain runtimes.

Reply:
Do we not need this excape hatch for the Mandatory inherents to avoid what happend on the rococo parachains?

Reply:
What happened on Rococo parachains ?

Reply:
It got stuck because an inherent ran a lazy migration that loaded a big Key pair and overflowed the PoV limit.

Reply:
@ggwpez Sorry for the late reply, I thought about the rococo situation before approval here.

So we don't have an explicit mechanism for mandatory extrinsics here because we will go into these two branches dring the weight checks:
1. This one because for mandatory `max_total` is `None`: https://github.com/paritytech/polkadot-sdk/blob/0bb931b6b1ba2f401a7f6af93f4cf885ef9e790e/substrate/frame/system/src/extensions/check_weight.rs#L206
2. This one because reserved is also `None` for mandatory extrinsics: https://github.com/paritytech/polkadot-sdk/blob/0bb931b6b1ba2f401a7f6af93f4cf885ef9e790e/substrate/frame/system/src/extensions/check_weight.rs#L222-L224

Of course what I described above is not guaranteed. Parachain teams can define in their runtime whatever they want. This is what the typical weight definition looks like (note no specifics defined for mandatory):
https://github.com/paritytech/polkadot-sdk/blob/0bb931b6b1ba2f401a7f6af93f4cf885ef9e790e/cumulus/parachains/runtimes/assets/asset-hub-rococo/src/lib.rs#L138-L155

The builder has `None` set by default, which makes sense:
https://github.com/paritytech/polkadot-sdk/blob/0bb931b6b1ba2f401a7f6af93f4cf885ef9e790e/substrate/frame/system/src/limits.rs#L347-L354

So with this PR the weight checking falls in line with the rest of the extrinsics. If someone has actually set a limit on mandatory weight class, they are living dangerous without the changes of this PR already. If some pallet returns max weight like the session pallet in the rococo case, we would already go over the limit together with `on_initialize`.